validator.go

```diff
@@ -30,61 +139,253 @@ func (v *Validator) Validate(schema *Schema, entity *egdm.Entity) (ok bool, exce
 		constraints := schema.GetConstraintsForEntityClass(class, true)
 		for _, constraint := range constraints {
 			// check if constraint is violated
-			if v.isViolated(constraint, entity) {
+			valid, violation, constraintError := v.CheckConstraint(schema, constraint, entity)
+			if constraintError != nil {
+				err = constraintError
+				return
+			}
+
+			if !valid && violation != nil {
 				ok = false
-				exceptions = append(exceptions, &ConstraintViolation{})
+				exceptions = append(exceptions, violation)
 			}
 		}
 	}
 
 	return
 }
 
-func (v *Validator) isViolated(constraint any, entity *egdm.Entity) bool {
+// CheckConstraint checks if the given constraint is violated for the given entity. Returns false if the constraint is ok
+// and true and a constraint violation struct if not. Error is returned if something went wrong while checking the constraint.
+func (v *Validator) CheckConstraint(schema *Schema, constraint any, entity *egdm.Entity) (bool, *ConstraintViolation, error) {
 	switch c := constraint.(type) {
 	case *ReferenceConstraint:
-		v.CheckReferenceConstraint(entity, c)
+		return v.CheckReferenceConstraint(entity, c)
 	case *InverseReferenceConstraint:
-		v.CheckInverseReferenceConstraint(entity, c)
+		return v.CheckInverseReferenceConstraint(entity, c)
 	case *PropertyConstraint:
-		v.CheckPropertyConstraint(entity, c)
+		return v.CheckPropertyConstraint(entity, c)
 	case *IsAbstractConstraint:
-		v.CheckIsAbstractConstraint(entity, c)
+		// this checks if this entity is an implementation of a EntityClass that is defined as abstract
+		return v.CheckEntityIsAbstractConstraint(schema, entity, c)
 	}
 
-	return false
+	return false, nil, errors.New("constraint type not supported")
 }
 
-type ValidationContext interface {
-	// return a given entity by id in the named datasets
-	GetEntity(id string, datasets []string) *egdm.Entity
+func (v *Validator) CheckReferenceConstraint(entity *egdm.Entity, constraint *ReferenceConstraint) (bool, *ConstraintViolation, error) {
```

Comment:
Cleaner to just return a Validation error, which is nil when ok and contains the violation object when raised?

Reply:
same as above.